### PR TITLE
Refactor Phoenix MERGE implementation

### PR DESCRIPTION
**Pull request by chenjian2664**
## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java [hunk 1]
```diff
@@ -64,90 +60,100 @@ public PhoenixMergeSink(PhoenixClient phoenixClient, RemoteQueryModifier remoteQ
...
+            String schemaName,
+            String tableName,
+            List<String> mergeRowIdFieldNames)
     {
-        public DeleteSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
-        {
-            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
-        }
+        List<String> conjuncts = mergeRowIdFieldNames.stream()
+                .map(name -> name + " = ? ")
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java:151]: I wonder if we should not use `columnWriters` from `JdbcPageSink` here. Notice that handling parameters is not trivial and I think we should somehow reuse it.

I need to think more how... any ideas?

Comment (chenjian2664) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java]: I am trying to use the `DefaultQueryBuilder#prepareDeleteQuery`, this is the code but not sure I am in the right direction.

Comment (chenjian2664) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java]: Hi, could you please help to have a look at the try. I am going to refactor the `Phoenix` until we meet the agreement, then I will modify the first commit of the #23034

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java [hunk 2]
```diff
@@ -111,10 +112,10 @@ public JdbcPageSink(ConnectorSession session, JdbcOutputTableHandle handle, Jdbc
                     .collect(toImmutableList());
         }
 
-        String sinkSql = getSinkSql(jdbcClient, handle, columnWriters);
         try {
-            sinkSql = remoteQueryModifier.apply(session, sinkSql);
-            statement = connection.prepareStatement(sinkSql);
+            String sql = sinkSql.orElseGet(() -> jdbcClient.buildInsertSql(handle, columnWriters));
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:116]: How about instead of injecting `Optional<String>` to inject `SinkSqlProvider`, where
```
interface SinkSqlProvider {
  String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters);
}
```

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java [hunk 3]
```diff
@@ -13,137 +13,74 @@
...
-import static org.apache.phoenix.util.SchemaUtil.getEscapedArgument;
 
 public class PhoenixMergeSink
         implements ConnectorMergeSink
 {
-    private final RemoteTableName remoteTableName;
-    private final boolean hasRowKey;
     private final int columnCount;
-    private final List<String> mergeRowIdFieldNames;
+    private final boolean hasRowKey;
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java:48]: nit: Maybe let's move it up, so the diff will be smaller

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java [hunk 4]
```diff
@@ -13,137 +13,74 @@
...
-                mergeRowIdFieldTypes,
-                Optional.empty(),
-                Optional.empty());
-
-        return new DeleteSink(session, deleteOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
-    }
-
-    private class DeleteSink
-            extends JdbcPageSink
+    public PhoenixMergeSink(PhoenixClient phoenixClient, RemoteQueryModifier remoteQueryModifier, ConnectorSession session, ConnectorMergeTableHandle connectorMergeTableHandle, ConnectorPageSinkId pageSinkId, QueryBuilder queryBuilder)
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java:54]: nit. Like above if that declaration would be divided into separate line it would be clear to see that you added one parameter at the end

If you decide to refactor these, please keep it in separate records

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeTableHandle.java [hunk 5]
```diff
@@ -15,42 +15,75 @@
...
 
     @JsonProperty
     @Override
-    public JdbcTableHandle getTableHandle()
+    public JdbcTableHandle tableHandle()
+    {
+        return tableHandle;
+    }
+
+    @Override
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeTableHandle.java:56]: Please add a comment why it is not `@JsonProperty`

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java [hunk 6]
```diff
@@ -13,137 +13,74 @@
...
+        PhoenixOutputTableHandle insertHandle = mergeTableHandle.insertHandle();
+
+        this.columnCount = insertHandle.getColumnTypes().size();
+        this.hasRowKey = insertHandle.rowkeyColumn().isPresent();
+
+        this.insertSink = new JdbcPageSink(session, insertHandle, phoenixClient, pageSinkId, remoteQueryModifier, Optional.empty());
+        this.updateSink = new JdbcPageSink(session, mergeTableHandle.updateHandle(), phoenixClient, pageSinkId, remoteQueryModifier, Optional.empty());
+
+        try (Connection connection = phoenixClient.getConnection(session)) {
+            String deleteSql = queryBuilder.prepareDeleteQuery(
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java:66]: Maybe usage of `queryBuilder` could be extracted to separate commit, to make it easier to follow the changes.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeTableHandle.java [hunk 7]
```diff
@@ -15,42 +15,75 @@
...
+import io.trino.spi.predicate.TupleDomain;
 
-public record PhoenixMergeTableHandle(JdbcTableHandle tableHandle, PhoenixOutputTableHandle phoenixOutputTableHandle, JdbcColumnHandle mergeRowIdColumnHandle)
+import static java.util.Objects.requireNonNull;
+
+public record PhoenixMergeTableHandle(
+        JdbcTableHandle tableHandle,
+        PhoenixOutputTableHandle insertHandle,
+        PhoenixOutputTableHandle updateHandle,
+        PhoenixOutputTableHandle deleteHandle,
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeTableHandle.java:30]: `PhoenixOutputTableHandle`s have similar data in it. I wonder if it is not a problem when serializing it that a lot of JSON will be actually duplicated.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java [hunk 8]
```diff
@@ -327,12 +332,44 @@ public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorT
...
+        // Delete only requires the primary keys
+        PhoenixOutputTableHandle deleteHandle = (PhoenixOutputTableHandle) beginInsert(session, plainTable, ImmutableList.copyOf(primaryKeys), retryMode);
+
+        // The TupleDomain for building the conjuncts of the primary keys
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `DefaultQueryBuilder#prepareDeleteQuery`
+        Domain dummy = Domain.singleValue(BIGINT, 0L);
+        for (JdbcColumnHandle columnHandle : primaryKeys) {
+            primaryKeysDomainBuilder.put(columnHandle, dummy);
+        }
```
Comment (kokosing) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java:365]: Why this code changed so much here?

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java (hunk 1 above)
Comment (kokosing) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java]: I think it is looking nice.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java (hunk 2 above)
Comment (chenjian2664) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:116]: Good idea! let me have a try~

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java (hunk 8 above)
Comment (chenjian2664) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java:365]: Prepared the `insert`, `update`, `delete` handle for the `PhoenixMergeSink` that can use directly. I will remove this change and to see if we actually need this later.

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
@hashhar @ebyhr any comments?

**Comment by kokosing** [comment]
Thank you!

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java [hunk 9]
```diff
@@ -337,10 +340,19 @@ public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorT
                 .collect(toImmutableList());
         PhoenixOutputTableHandle phoenixOutputTableHandle = (PhoenixOutputTableHandle) beginInsert(session, plainTable, ImmutableList.copyOf(columns), retryMode);
 
+        // The TupleDomain for building the conjuncts of the primary keys
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `DefaultQueryBuilder#prepareDeleteQuery`
+        Domain dummy = Domain.singleValue(BIGINT, 0L);
+        for (JdbcColumnHandle columnHandle : phoenixClient.getPrimaryKeyColumnHandles(session, plainTable)) {
+            primaryKeysDomainBuilder.put(columnHandle, dummy);
+        }
```
Comment (hashhar) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java:349]: this is the only thing that's a bit weird to me.

I don't understand why we say "it won't affect query".

The `DefaultQueryBuilder` can change in future. Either way it does use the tuple domain to create conjuncts by passing this to `toConjuncts` method.

**Review by hashhar — COMMENTED** [review]
Nice work on this.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java (hunk 9 above)
Comment (chenjian2664) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java:349]: Thanks for the review.
> I don't understand why we say "it won't affect query".
> 
The domain does not participate in building the conjuncts as long as it is the singValue domain, this is what I want, only use the column and bind exception. It's a bit awkward to reuse the `DefaultQueryBuilder`, but currently the `DefaultQueryBuilder` not support function only accept the `column handles` to only get sql string.

> The `DefaultQueryBuilder` can change in future. Either way it does use the tuple domain to create conjuncts by passing this to `toConjuncts` method.

You are right, so after the merge got accept, the `DefaultQueryBuilder` needs to keep the correctness of the result that with the parameters: column & singDomain.
We need either implement a new query builder or provide/extract a new function to support the current situation to let the `QueryBuilder` more easily to maintain, I can have a try on it.:)

Comment (hashhar) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java:349]: Thanks for the explanation.

> provide/extract a new function to support the current situation to let the QueryBuilder more easily to maintain, I can have a try on it.:)

I think this makes sense to do.
